File: agentwallet/packages/sdk-python/src/agentwallet/resources/transactions.py
"""Transactions sub-resource."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..types import ListResponse, Transaction

if TYPE_CHECKING:
    from ..client import AgentWallet


class TransactionsResource:
    def __init__(self, client: AgentWallet):
        self._client = client

    async def transfer_sol(
        self,
        from_wallet: str,
        to_address: str,
        amount_sol: float,
        memo: str | None = None,
        idempotency_key: str | None = None,
    ) -> Transaction:
        data = await self._client.post("/transactions/transfer-sol", json={
            "from_wallet_id": from_wallet,
            "to_address": to_address,
            "amount_sol": amount_sol,
            "memo": memo,
            "idempotency_key": idempotency_key,
        })
        return Transaction(**data)

    async def get(self, tx_id: str) -> Transaction:
        data = await self._client.get(f"/transactions/{tx_id}")
        return Transaction(**data)

    async def list(
        self,
        agent_id: str | None = None,
        wallet_id: str | None = None,
        status: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> ListResponse:
        params = {"limit": limit, "offset": offset}
        if agent_id:
            params["agent_id"] = agent_id
        if wallet_id:
            params["wallet_id"] = wallet_id
        if status:
            params["status"] = status
        data = await self._client.get("/transactions", params=params)
        return ListResponse(
            data=[Transaction(**t) for t in data["data"]],
            total=data["total"],
        )

    async def batch_transfer(
        self, transfers: list[dict]
    ) -> list[Transaction]:
        data = await self._client.post("/transactions/batch-transfer", json={
            "transfers": transfers,
        })
        return [Transaction(**t) for t in data]

File: agentwallet/packages/sdk-ts/src/resources/transactions.ts
import type { AgentWallet } from '../client';
import type { BatchTransferItem, ListResponse, Transaction, TransferSolParams } from '../types';

export class TransactionsResource {
  constructor(private client: AgentWallet) {}

  /** Transfer SOL between wallets */
  async transferSol(params: TransferSolParams): Promise<Transaction> {
    return this.client.post<Transaction>('/transactions/transfer-sol', params as unknown as Record<string, unknown>);
  }

  /** Get a transaction by ID */
  async get(txId: string): Promise<Transaction> {
    return this.client.get<Transaction>(`/transactions/${txId}`);
  }

  /** List transactions */
  async list(options?: {
    agent_id?: string;
    wallet_id?: string;
    status?: string;
    limit?: number;
    offset?: number;
  }): Promise<ListResponse<Transaction>> {
    return this.client.get<ListResponse<Transaction>>('/transactions', {
      agent_id: options?.agent_id,
      wallet_id: options?.wallet_id,
      status: options?.status,
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    });
  }

  /** Batch transfer SOL to multiple recipients */
  async batchTransfer(transfers: BatchTransferItem[]): Promise<Transaction[]> {
    return this.client.post<Transaction[]>('/transactions/batch-transfer', {
      transfers,
    } as unknown as Record<string, unknown>);
  }
}
